scanpy/preprocessing/_utils.py

```diff
@@ -1,29 +1,127 @@
 import numpy as np
-from scipy.sparse import issparse
+from scipy import sparse
+import numba
 
 
-STANDARD_SCALER_FIXED = False
+def _get_mean_var(X):
+    if sparse.issparse(X):
+        mean, var = sparse_mean_variance_axis(X, axis=0)
+    else:
+        mean = np.mean(X, axis=0, dtype=np.float64)
+        mean_sq = np.multiply(X, X).mean(axis=0, dtype=np.float64)
+        var = mean_sq - mean ** 2
+    # enforce R convention (unbiased estimator) for variance
+    var *= X.shape[0] / (X.shape[0] - 1)
+    return mean, var
 
 
-def _get_mean_var(X):
-    # - using sklearn.StandardScaler throws an error related to
-    #   int to long trafo for very large matrices
-    # - using X.multiply is slower
-    if not STANDARD_SCALER_FIXED:
-        mean = X.mean(axis=0)
-        if issparse(X):
-            mean_sq = X.multiply(X).mean(axis=0)
-            mean = mean.A1
-            mean_sq = mean_sq.A1
-        else:
-            mean_sq = np.multiply(X, X).mean(axis=0)
-        # enforece R convention (unbiased estimator) for variance
-        var = (mean_sq - mean ** 2) * (X.shape[0] / (X.shape[0] - 1))
+def sparse_mean_variance_axis(mtx: sparse.spmatrix, axis: int):
+    """
+    This code and internal functions are based on sklearns
+    `sparsefuncs.mean_variance_axis`.
+
+    Modifications:
+    * allow deciding on the output type, which can increase accuracy when calculating the mean and variance of 32bit floats.
+    * This doesn't currently implement support for null values, but could.
+    * Uses numba not cython
+    """
+    assert axis in (0, 1)
+    if isinstance(mtx, sparse.csr_matrix):
+        if axis == 0:
+            return sparse_mean_var_minor_axis(
+                mtx.data, mtx.indices, mtx.shape[0], mtx.shape[1], np.float64
+            )
+        elif axis == 1:
+            return sparse_mean_var_major_axis(
+                mtx.data,
+                mtx.indices,
+                mtx.indptr,
+                mtx.shape[0],
+                mtx.shape[1],
+                np.float64,
+            )
+    elif isinstance(mtx, sparse.csc_matrix):
+        if axis == 0:
+            return sparse_mean_var_major_axis(
+                mtx.data,
+                mtx.indices,
+                mtx.indptr,
+                mtx.shape[1],
+                mtx.shape[0],
+                np.float64,
+            )
+        elif axis == 1:
+            return sparse_mean_var_minor_axis(
+                mtx.data, mtx.indices, mtx.shape[1], mtx.shape[0], np.float64
+            )
     else:
-        from sklearn.preprocessing import StandardScaler
+        raise ValueError(
+            "This function only works on sparse csr and csc matrices"
+        )
```

Comment:
please condense this to two branches instead of 4.

```py
csr = isinstance(mtx, sparse.csr_matrix)
if not csr and not isinstance(mtx, sparse.csc_matrix):
    raise ValueError(
        "This function only works on sparse csr and csc matrices"
    )
shape = mtx.shape if csr else mtx.shape[::-1]
if axis == int(csr):
    return sparse_mean_var_major_axis(
        mtx.data, mtx.indices, mtx.indptr, *shape, np.float64
    )
else:
    return sparse_mean_var_minor_axis(
        mtx.data, mtx.indices, *shape, np.float64
    )
```

Reply:
I'm not a fan of `if axis == int(csr)`. For one thing it would have to be rewritten for COO. Maybe this?

```python
@singledispatch
def sparse_mean_variance_axis(mtx, axis):
    """
    This code and internal functions are based on sklearns
    `sparsefuncs.mean_variance_axis`.

    Modifications:
    * allow deciding on the output type, which can increase accuracy when
    calculating the mean and variance of 32bit floats.
    * This doesn't currently implement support for null values, but could.
    * Uses numba not cython
    """
    raise NotImplementedError(
        "This function is only defined for sparse csr and csc matrices"
    )


@sparse_mean_variance_axis.register(sparse.csr_matrix)
def sparse_mean_variance_axis_csr(mtx, axis):
    assert axis in (0, 1)
    if axis == 0:
        return sparse_mean_var_minor_axis(
            mtx.data, mtx.indices, *mtx.shape, np.float64
        )
    elif axis == 1:
        return sparse_mean_var_major_axis(
            mtx.data, mtx.indices, mtx.indptr, *mtx.shape, np.float64
        )


@sparse_mean_variance_axis.register(sparse.csc_matrix)
def sparse_mean_variance_axis_csc(mtx, axis):
    assert axis in (0, 1)
    if axis == 0:
        return sparse_mean_var_major_axis(
            mtx.data, mtx.indices, mtx.indptr, *mtx.shape[::-1], np.float64
        )
    elif axis == 1:
        return sparse_mean_var_minor_axis(
            mtx.data, mtx.indices, *mtx.shape[::-1], np.float64
        )
```

It's also purely internal code, so I'm  not sure it's worth spending too much time on style.

Reply:
My code is the complete thing, so nothing has to be rewritten.

If you think my version is too cute, you can use better variable names or as above an `if else` expression instead of converting the bool to an int.

I’m also OK with your version if you replace the last function with:

```py
@sparse_mean_variance_axis.register(sparse.csc_matrix)
def sparse_mean_variance_axis_csc(mtx, axis):
    # use csr implementation
    return sparse_mean_variance_axis(mtx.T, 1 - axis)
```

Reply:
Oh, I meant to say a boolean switch stops being sufficient if there are 3 options. `i.e` is sparse matrices could be `scipy`'s CSR and CSC or `pydata/sparse`'s COO. If COO is introduced the logic has to be redone, as opposed to just adding another case.

In general I prefer the original, since the logic is pretty simple. I think [this blog post](https://mrocklin.github.com/blog/work/2019/06/23/avoid-indirection) is a good discussion of this idea, as is the one linked at the bottom. As you can probably tell, I generally like fancy code, but the tool has to fit the problem.

I tried `mtx.T` before, but it was a bit slower. `scipy.sparse` can be slow to construct sparse matrices for some reason (not sure why, since it doesn't do much validation).

Also, would it be `abs(axis - 1)` not `1 - axis`?

Reply:
Yeah, it's always get's me turned around a bit that `axis` is the one being reduced, so you get the means for the other axes. I think referring to major and minor is a better convention than the original sklearn names though.